File: samples/node-billing-svc/src/service/BillingService.js
import { CustomerRepository } from '../repo/CustomerRepository.js';
import { StripePaymentGateway } from '../gateway/StripePaymentGateway.js';
import { EventBus } from '../events/EventBus.js';

function isUuidLike(s) { return typeof s === 'string' && s.length >= 8; }
function isEmailLike(s) { return typeof s === 'string' && s.includes('@'); }

export async function handleUserCreated(eventDTO) {
  // validation
  if (!eventDTO || typeof eventDTO !== 'object') {
    return Object.freeze({ status: 'ignored' });
  }
  const { userId, email, displayName } = eventDTO;
  if (!isUuidLike(userId) || !isEmailLike(email)) {
    return Object.freeze({ status: 'ignored' });
  }

  // decision: already exists?
  if (CustomerRepository.existsById(userId)) {
    return Object.freeze({ status: 'ignored' });
  }

  // domain-service-call: create stripe customer
  const stripeCustomer = StripePaymentGateway.createCustomer(email, displayName ?? '');

  // mapping to CustomerEntity
  const newCustomer = Object.freeze({
    customerId: String(userId),
    email: String(email),
    displayName: displayName ?? '',
    stripeCustomerId: stripeCustomer.id,
  });

  // persistence-call
  const savedCustomer = CustomerRepository.save(newCustomer);

  // publish-event: CustomerCreatedEvent
  EventBus.publish('CustomerCreatedEvent', savedCustomer);

  // return
  return Object.freeze({ status: 'created' });
}

export const BillingService = Object.freeze({ handleUserCreated });

File: samples/billing-svc/src/main/java/com/example/billing/service/BillingService.java
package com.example.billing.service;

import com.example.billing.domain.CustomerEntity;
import com.example.billing.dto.UserCreatedEventDTO;
import com.example.billing.events.CustomerCreatedEvent;
import com.example.billing.gateway.PaymentGateway;
import com.example.billing.repo.CustomerRepository;
import org.springframework.context.ApplicationEventPublisher;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Map;
import java.util.UUID;

@Service
public final class BillingService {

    private final CustomerRepository customerRepository;
    private final PaymentGateway paymentGateway;
    private final CustomerFactory customerFactory;
    private final ApplicationEventPublisher eventPublisher;

    public BillingService(final CustomerRepository customerRepository,
                          final PaymentGateway paymentGateway,
                          final CustomerFactory customerFactory,
                          final ApplicationEventPublisher eventPublisher) {
        this.customerRepository = customerRepository;
        this.paymentGateway = paymentGateway;
        this.customerFactory = customerFactory;
        this.eventPublisher = eventPublisher;
    }

    @Transactional
    public Map<String, String> handleUserCreated(final UserCreatedEventDTO eventDTO) {
        // validation (basic): userId UUID and email contains @
        try { UUID.fromString(eventDTO.userId()); } catch (Exception e) { throw new IllegalArgumentException("userId must be UUID"); }
        if (eventDTO.email() == null || !eventDTO.email().contains("@")) {
            throw new IllegalArgumentException("email must be valid");
        }

        final UUID userId = UUID.fromString(eventDTO.userId());

        // exists check
        if (customerRepository.existsByCustomerId(userId)) {
            return Map.of("status", "ignored");
        }

        // Adapter pattern: create customer in Stripe via PaymentGateway
        final Map<String, Object> stripeCustomer = paymentGateway.createCustomer(eventDTO.email(), eventDTO.displayName());
        final String stripeId = (String) stripeCustomer.get("id");

        // Factory pattern: create aggregate from event
        final CustomerEntity newCustomer = customerFactory.createCustomerFromEvent(eventDTO, stripeId);

        // persist
        final CustomerEntity saved = customerRepository.save(newCustomer);

        // publish event
        eventPublisher.publishEvent(new CustomerCreatedEvent(this, saved.getCustomerId(), saved.getEmail()));

        return Map.of("status", "created");
    }
}
